README.md
# Timber.io - Master your Go apps with structured logging

Timber Go is under development, currently this library only contains the core components to ship logs to the Timber API.

## Packages

### `batch`

An implementation for efficiently collecting and sending strings via input and output channels. Batches are sent when the configured buffer size is exceeded or if the configured time has elasped since the last send.

### `forward`

Exposes an a Forwarder interface for accepting a buffer and writing it somewhere. Implementations include stdout, file,
and http forwarders.

### `logging`

Exposes a logger interface that all timber-go packages and their dependencies use. Aside from bring your own, there is
the default logger to stdout and the discard logger to /dev/null. This allows for deciding how library logs are handled
within their dependents.

### `metadata`

An implementation of Timber's metadata JSON schema. It also includes metadata collection for the supported platforms
(e.g. AWS).

## LICENSE

The original parts of this software as developed by Timber Technologies, Inc. as
well as contributors are licensed under the Internet Systems Consortium (ISC)
License. This software is dependent upon third-party code which is
statically linked into the executable at compile time. This third-party code is
redistributed without modification and made available to all users  under the
terms of each project's original license within the `vendor` directory of the
project.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.1.0] - 2018-06-06

### Changed

  - Use `[]byte` instead of `string` for batcher input to avoid unnecessary overhead.
  - Discard httpClient logs as they were noisy and redundant.

## [0.0.1] - 2018-05-11

### Added

  - Imported core functionality for handling logs in memory from [Timber Agent](https://github.com/timberio/agent)

[Unreleased]: https://github.com/timberio/timber-go/compare/v0.1.0...HEAD
[0.1.0]: https://github.com/timberio/timber-go/compare/v0.0.1...v0.1.0
[0.0.1]: https://github.com/timberio/timber-go/tree/v0.0.1


batch/batching_test.go
package batch

import (
	"bytes"
	"testing"
	"time"
)

func TestChannelClosing(t *testing.T) {
	byteChan := make(chan []byte)

	batcher := NewBatcher(byteChan, Config{
		Period: 10 * time.Second,
	})

	batcher.ByteChan <- []byte("test log line")
	close(batcher.ByteChan)

	actual := <-batcher.BufferChan
	expected := "test log line\n"
	if actual.String() != expected {
		t.Fatalf("expected \"%+v\", got \"%+v\"", expected, actual)
	}
}

func TestBufferOverflow(t *testing.T) {
	byteChan := make(chan []byte)
	batcher := NewBatcher(byteChan, DefaultConfig())

	filler := []byte("test log line")
	fillerLen := len(filler) + 1
	for written := 0; written+fillerLen < 990000; written += fillerLen {
		batcher.ByteChan <- filler
	}
	batcher.ByteChan <- []byte("overflowed")
	close(batcher.ByteChan)

	<-batcher.BufferChan // throw away the big one
	actual := <-batcher.BufferChan
	expected := "overflowed\n"
	if actual.String() != expected {
		t.Fatalf("expected \"%+v\", got \"%+v\"", expected, actual)
	}
}

// Batch()
// Log lines larger than the max payload size (1 MB) should be dropped
func TestBatchDropLogLine(t *testing.T) {
	byteChan := make(chan []byte)
	batcher := NewBatcher(byteChan, DefaultConfig())

	filler := "test log line"
	buf := bytes.NewBuffer(make([]byte, defaultBufferSize))
	for buf.Len() < defaultBufferSize {
		buf.WriteString(filler)
	}
	logline := buf.String()

	// go Batch(lines, bufChan, 10)
	batcher.ByteChan <- []byte(logline)
	close(batcher.ByteChan)

	// Nothing should be sent to bufChan since we are dropping message
	actual := <-batcher.BufferChan
	if actual != nil {
		t.Fatalf("expected \"%+v\" to be nil", actual)
	}
}


forward/forwarding_test.go
package forward

import (
	"bytes"
	"io/ioutil"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
)

func TestForwardForwarding(test *testing.T) {
	bufChan := make(chan *bytes.Buffer, 1)
	bufChan <- bytes.NewBufferString("test log line\n")
	close(bufChan)

	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(200)

		output, err := ioutil.ReadAll(r.Body)
		if err != nil {
			test.Fatal(err)
		}
		actual := strings.TrimSpace(bytes.NewBuffer(output).String())
		expected := "test log line"

		if actual != expected {
			test.Fatalf("expected \"%+v\", got \"%+v\"", expected, actual)
		}
	}))
	defer ts.Close()

	httpForwarder, _ := NewHTTPForwarder("api key", DefaultConfig())
	Forward(bufChan, httpForwarder)
}

func TestForwardRetries(test *testing.T) {
	bufChan := make(chan *bytes.Buffer, 1)
	bufChan <- bytes.NewBufferString("test log line\n")
	close(bufChan)

	retries := 0
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if retries < 1 {
			retries += 1
			w.WriteHeader(500)
		} else {
			w.WriteHeader(200)
		}
	}))
	defer ts.Close()

	httpForwarder, _ := NewHTTPForwarder("api key", Config{
		Endpoint: ts.URL,
	})
	httpForwarder.HTTPClient.RetryWaitMin = 0
	Forward(bufChan, httpForwarder)

	if retries != 1 {
		test.Fatalf("expected 1 retry, got %d", retries)
	}
}

func TestForwardMetadata(test *testing.T) {
	bufChan := make(chan *bytes.Buffer, 1)
	bufChan <- bytes.NewBufferString("test log line\n")
	close(bufChan)

	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		expected := "Metadata test"
		actual := r.Header.Get("Timber-Metadata-Override")

		if actual != expected {
			test.Fatalf("expected \"%+v\", got \"%+v\"", expected, actual)
		}

		w.WriteHeader(200)
	}))

	defer ts.Close()

	httpForwarder, _ := NewHTTPForwarder("api key", Config{
		Endpoint: ts.URL,
		Metadata: "Metadata test",
	})
	Forward(bufChan, httpForwarder)
}


metadata/log_event_test.go
package metadata

import (
	"testing"
)

// NOTE: These tests are not the best they could be since they expect the JSON to always
// be built the same way. Since JSON is order-independent, that's a bad expectation. This
// works for now, but a better solution is needed in the future.

// The following tests building a log event with system context; the intent of this test is to make sure that the
// JSON payload does not carry the "empty" child attributes of event.Context (event.Context.Platform and
// event.Context.Source).
func TestLogEventEncodeJSONEmptyChild(test *testing.T) {
	expected := `{"$schema":"https://raw.githubusercontent.com/timberio/log-event-json-schema/v3.0.8/schema.json","context":{"system":{"hostname":"localhost"}}}`
	event := NewLogEvent()
	event.ensureSystemContext()
	event.Context.System.Hostname = "localhost"
	encodedBytes, err := event.EncodeJSON()

	if err != nil {
		test.Fatal("Could not encode the event as JSON!")
	}

	encodedString := string(encodedBytes)

	if encodedString != expected {
		test.Fatalf("Expected %s but got %s", expected, encodedString)
	}
}


batch/batching.go
package batch

import (
	"bytes"
	"log"
	"os"
	"time"

	"github.com/timberio/timber-go/logging"
)

var (
	// Preallocate 990kb. The Timber API will not accept payloads larger than 1mb.
	// This leaves 10kb for headers.
	defaultBufferSize = 990000

	defaultPeriod = 3 * time.Second
)

type Batcher struct {
	BufferChan chan *bytes.Buffer
	ByteChan   chan []byte

	Config
}

type Config struct {
	Period time.Duration
	Size   int

	Logger logging.Logger
}

func DefaultConfig() Config {
	return Config{
		Period: defaultPeriod,
		Size:   defaultBufferSize,

		Logger: log.New(os.Stderr, "", log.LstdFlags),
	}
}

func NewBatcher(byteChan chan []byte, config Config) *Batcher {
	defaultConfig := DefaultConfig()

	if config.Period == 0 {
		config.Period = defaultConfig.Period
	}

	if config.Size == 0 {
		config.Size = defaultConfig.Size
	}

	if config.Logger == nil {
		config.Logger = defaultConfig.Logger
	}

	batcher := &Batcher{
		BufferChan: make(chan *bytes.Buffer),
		ByteChan:   byteChan,
		Config:     config,
	}

	go batcher.batch()

	return batcher
}

func Batch(byteChan chan []byte) *Batcher {
	batcher := &Batcher{
		BufferChan: make(chan *bytes.Buffer),
		ByteChan:   byteChan,
		Config:     DefaultConfig(),
	}

	go batcher.batch()

	return batcher
}

func (batcher *Batcher) batch() {
	buffer := freshBuffer(batcher.Size)
	tick := time.Tick(batcher.Period)

	for {
		select {
		case b, ok := <-batcher.ByteChan:
			if ok {
				if len(b)+1 > buffer.Cap() {
					// @TODO track souce?
					// @TODO more informative log line
					// have callback channel for results
					batcher.Logger.Print("Dropping log line greater than the max buffer size")
					continue
				}

				if buffer.Len()+len(b)+1 > buffer.Cap() {
					batcher.BufferChan <- buffer
					buffer = freshBuffer(batcher.Size)
				}

				if len(b) > 0 {
					b = append(b, []byte("\n")...)
					buffer.Write(b)
				}

			} else { // channel is closed
				if buffer.Len() > 0 {
					batcher.BufferChan <- buffer
				}
				close(batcher.BufferChan)
				return
			}

		case <-tick:
			if buffer.Len() > 0 {
				batcher.BufferChan <- buffer
				buffer = freshBuffer(batcher.Size)
			}
		}
	}
}

func freshBuffer(size int) *bytes.Buffer {
	buf := bytes.NewBuffer(make([]byte, size))
	buf.Reset()
	return buf
}


metadata/log_event.go
package metadata

import (
	"encoding/json"
)

var schema string = "https://raw.githubusercontent.com/timberio/log-event-json-schema/v3.0.8/schema.json"

type LogEvent struct {
	Schema  string   `json:"$schema"`
	Context *Context `json:"context,omitempty"`
}

type Context struct {
	System   *SystemContext   `json:"system,omitempty"`
	Platform *PlatformContext `json:"platform,omitempty"`
	Source   *SourceContext   `json:"source,omitempty"`
}

type SystemContext struct {
	Hostname string `json:"hostname,omitempty"`
}

type PlatformContext struct {
	AWSEC2 *AWSEC2Context `json:"aws_ec2,omitempty"`
}

type SourceContext struct {
	FileName string `json:"file_name,omitempty"`
}

type AWSEC2Context struct {
	AmiID          string `json:"ami_id,omitempty"`
	Hostname       string `json:"hostname,omitempty"`
	InstanceID     string `json:"instance_id,omitempty"`
	InstanceType   string `json:"instance_type,omitempty"`
	PublicHostname string `json:"public_hostname,omitempty"`
}

func NewLogEvent() *LogEvent {
	return &LogEvent{Schema: schema}
}

func (logEvent *LogEvent) AddEC2Context(context *AWSEC2Context) {
	logEvent.ensurePlatformContext()
	logEvent.Context.Platform.AWSEC2 = context
}

func (logEvent *LogEvent) AddHostname(hostname string) {
	logEvent.ensureSystemContext()
	logEvent.Context.System.Hostname = hostname
}

func (logEvent *LogEvent) EncodeJSON() ([]byte, error) {
	return json.Marshal(logEvent)
}

func (logEvent *LogEvent) ensureContext() {
	if logEvent.Context == nil {
		logEvent.Context = &Context{}
	}
}

func (logEvent *LogEvent) ensurePlatformContext() {
	logEvent.ensureContext()
	if logEvent.Context.Platform == nil {
		logEvent.Context.Platform = &PlatformContext{}
	}
}

func (logEvent *LogEvent) ensureSystemContext() {
	logEvent.ensureContext()
	if logEvent.Context.System == nil {
		logEvent.Context.System = &SystemContext{}
	}
}

func (logEvent *LogEvent) ensureSourceContext() {
	logEvent.ensureContext()
	if logEvent.Context.Source == nil {
		logEvent.Context.Source = &SourceContext{}
	}
}


forward/forwarder_file.go
package forward

import (
	"bufio"
	"bytes"
	"os"
	"path/filepath"

	"github.com/timberio/timber-go/logging"
)

type FileForwarder struct {
	Filename string
	Logger   logging.Logger

	file *os.File
}

func NewFileForwarder(filename string, logger logging.Logger) (*FileForwarder, error) {
	if logger == nil {
		logger = logging.DefaultLogger
	}

	file, err := openFile(filename)
	if err != nil {
		return nil, err
	}

	return &FileForwarder{
		Filename: filename,

		file: file,
	}, nil
}

func openFile(filename string) (*os.File, error) {
	err := os.MkdirAll(filepath.Dir(filename), os.ModePerm)
	if err != nil {
		return nil, err
	}

	var file *os.File

	if _, err := os.Stat(filename); err == nil {
		file, err = os.OpenFile(filename, os.O_RDWR, os.ModePerm)
		if err != nil {
			return nil, err
		}
	} else {
		file, err = os.Create(filename)
		if err != nil {
			return nil, err
		}
	}

	return file, nil
}

func (f *FileForwarder) Forward(buffer *bytes.Buffer) {
	writer := bufio.NewWriter(f.file)

	_, err := writer.Write(buffer.Bytes())
	if err != nil {
		f.Logger.Print(err)
		return
	}

	err = writer.Flush()
	if err != nil {
		f.Logger.Print(err)
		return
	}
}


forward/forwarder_http.go
package forward

import (
	"bytes"
	"encoding/base64"
	"errors"
	"fmt"
	"time"

	"github.com/hashicorp/go-retryablehttp"
	"github.com/timberio/timber-go/logging"
)

var (
	// Replaced at compile time
	version = "undefined"

	defaultHTTPForwarderEndpoint = "https://logs.timber.io/frames"
	defaultHTTPForwarderTimeout  = 10 * time.Second
)

type HTTPForwarder struct {
	HTTPClient *retryablehttp.Client
	APIKey     string

	Config
}

type Config struct {
	Endpoint  string
	Metadata  string
	UserAgent string

	Logger logging.Logger
}

func DefaultConfig() Config {
	return Config{
		Endpoint:  defaultHTTPForwarderEndpoint,
		UserAgent: fmt.Sprintf("timber-go-forward/%s", version),

		Logger: logging.DefaultLogger,
	}
}

func NewHTTPForwarder(apiKey string, config Config) (*HTTPForwarder, error) {
	if apiKey == "" {
		return nil, errors.New("API KEY REQUIRED")
	}

	defaultConfig := DefaultConfig()

	if config.Endpoint == "" {
		config.Endpoint = defaultConfig.Endpoint
	}

	if config.UserAgent == "" {
		config.UserAgent = defaultConfig.UserAgent
	}

	if config.Logger == nil {
		config.Logger = defaultConfig.Logger
	}

	httpClient := retryablehttp.NewClient()
	httpClient.Logger = logging.DiscardingLogger
	httpClient.HTTPClient.Timeout = defaultHTTPForwarderTimeout

	return &HTTPForwarder{
		HTTPClient: httpClient,
		APIKey:     apiKey,
		Config:     config,
	}, nil
}

func (h *HTTPForwarder) Forward(buffer *bytes.Buffer) {
	token := base64.StdEncoding.EncodeToString([]byte(h.APIKey))
	authorization := fmt.Sprintf("Basic %s", token)

	req, err := retryablehttp.NewRequest("POST", h.Endpoint, bytes.NewReader(buffer.Bytes()))
	if err != nil {
		h.Logger.Print(err)
		return
	}

	req.Header.Add("Content-Type", "text/plain")
	req.Header.Add("Authorization", authorization)
	req.Header.Add("User-Agent", h.UserAgent)
	req.Header.Add("Timber-Metadata-Override", h.Metadata)

	resp, err := h.HTTPClient.Do(req)
	if err != nil {
		// retries have already happened at this point, so give up
		h.Logger.Print(err)
		return
	}
	resp.Body.Close()

	if resp.StatusCode >= 300 {
		//@TODO: More informative logging message
		h.Logger.Printf("HTTPForwarder: unexpected response (status code %d)", resp.StatusCode)
		return
	}
}


metadata/aws_ec2_metadata.go
package metadata

import (
	"errors"
	"io/ioutil"
	"net/http"
	"time"

	"github.com/timberio/timber-go/logging"
)

var (
	defaultTimeout  = 1 * time.Second
	defaultEndpoint = "http://169.254.169.254"
)

type Config struct {
	Timeout time.Duration

	Logger logging.Logger
}

type EC2Client struct {
	BaseEndpoint string
	HTTPClient   *http.Client

	Config
}

func DefaultConfig() Config {
	return Config{
		Timeout: defaultTimeout,
		Logger:  logging.DefaultLogger,
	}
}

func NewEC2Client(config Config) *EC2Client {
	defaultConfig := DefaultConfig()

	if config.Timeout == 0 {
		config.Timeout = defaultConfig.Timeout
	}

	if config.Logger == nil {
		config.Logger = defaultConfig.Logger
	}

	client := &http.Client{
		Timeout: config.Timeout,
	}

	return &EC2Client{
		BaseEndpoint: defaultEndpoint,
		HTTPClient:   client,
		Config:       config,
	}
}

func AddEC2Metadata(client *EC2Client, logEvent *LogEvent) {
	if !client.Available() {
		client.Logger.Print("Agent is not running on an EC2 instance")
		return
	} else {
		client.Logger.Print("Agent is running on an EC2 instance")
	}

	amiID, err := client.GetMetadata("ami-id")

	if err != nil {
		client.Logger.Print("Could not determine AMI ID the EC2 instance was launched with")
	} else {
		client.Logger.Printf("Discovered AMI ID from AWS EC2 metadata: %s", amiID)
	}

	hostname, err := client.GetMetadata("hostname")

	if err != nil {
		client.Logger.Print("Cloud not determine the AWS assigned hostname for the EC2 instance")
	} else {
		client.Logger.Printf("Discovered EC2 Hostname from AWS EC2 metadata: %s", hostname)
	}

	instanceID, err := client.GetMetadata("instance-id")

	if err != nil {
		client.Logger.Print("Could not determine the instance ID for the EC2 instance")
	} else {
		client.Logger.Printf("Discovered Instance ID from AWS EC2 metadata: %s", instanceID)
	}

	instanceType, err := client.GetMetadata("instance-type")

	if err != nil {
		client.Logger.Print("Could not determine the instance type for the EC2 instance")
	} else {
		client.Logger.Printf("Discovered Instance Type from AWS EC2 metadata: %s", instanceType)
	}

	publicHostname, err := client.GetMetadata("public-hostname")

	if err != nil {
		client.Logger.Print("Could not determine the AWS assigned public hostname for the EC2 instance")
	} else {
		client.Logger.Printf("Discovered EC2 Public Hostname from AWS EC2 metadata: %s", publicHostname)
	}

	context := &AWSEC2Context{
		AmiID:          amiID,
		Hostname:       hostname,
		InstanceID:     instanceID,
		InstanceType:   instanceType,
		PublicHostname: publicHostname,
	}

	logEvent.AddEC2Context(context)
}

func (client *EC2Client) Available() bool {
	resp, err := client.HTTPClient.Get(client.BaseEndpoint + "/latest/meta-data/")

	if err != nil {
		return false
	}

	resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return false
	}

	return true
}

func (client *EC2Client) GetMetadata(field string) (string, error) {
	resp, err := client.HTTPClient.Get(client.BaseEndpoint + "/latest/meta-data/" + field)

	if err != nil {
		return "", err
	}

	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return "", errors.New("Did not received a valid response for EC2 metadata")
	}

	body, err := ioutil.ReadAll(resp.Body)

	if err != nil {
		return "", err
	}

	return string(body[:]), nil
}
